go-controller/pkg/clustermanager/dnsnameresolver/controller.go

```diff
@@ -0,0 +1,541 @@
+	}
+	// Fetch the dns name resolver object using the name and namespace.
+	obj, err := c.dnsLister.DNSNameResolvers(namespace).Get(name)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			// DNSNameResolver object was deleted. If it is not supposed to be deleted,
+			// recreate it.
+
+			// Check if the details of the object and the corresponding DNS
+			// name exists or not.
+			dnsName, found := c.objDNSName[name]
+			if !found {
+				return nil
+			}
+			// Get the object details corresponding to the DNS name.
+			objDetails, exists := c.dnsNameObj[dnsName]
+			if !exists {
+				// If the object details doesn't exists in the dnsNameObj map,
+				// then remove the corresponding entry from the objDNSName map.
+				delete(c.objDNSName, name)
+				return nil
+			}
+			// Check if the DNS name is used in any of the namespaces. If not,
+			// then remove the details of the DNS name and the object.
+			if objDetails.namespaces.Len() == 0 {
+				delete(c.dnsNameObj, dnsName)
+				delete(c.objDNSName, objDetails.objName)
+				return nil
+			}
+
+			// Recreate the DNSNameResolver object.
+			klog.Warningf("Recreating deleted dns name resolver object %s for dns name %s", name, dnsName)
+			return c.createDNSNameResolver(name, dnsName)
+
+		}
+		return fmt.Errorf("reconcileDNSNameResolver failed to fetch dns name resolver %s in namespace %s", name, namespace)
+	}
+
+	// DNSNameResolver object was added/updated. If it is not supposed to exist,
+	// delete the object.
+
+	dnsName := string(obj.Spec.Name)
+	if objDetails, exists := c.dnsNameObj[dnsName]; !exists || objDetails.objName != obj.Name {
+		// Delete the DNSNameResolver object.
+		klog.Warningf("Deleting additional dns name resolver object %s for dns name %s", obj.Name, dnsName)
+		return c.deleteDNSNameResolver(obj.Name)
+	}
+	return nil
+}
+
+// addDNSName creates the corresponding DNSNameResolver object for the DNS name
+// if not already created. The namespace is added to the list of namespaces where
+// the DNS name is used.
+func (c *Controller) addDNSName(dnsName, namespace string) error {
+	objDetails, exists := c.dnsNameObj[dnsName]
+	// Create the DNSNameResolver object if its details are not available.
+	if !exists || objDetails.objName == "" {
+		// The collisionCount variable for the DNS name is incremented a maximum
+		// of 10 times for generating a name for the corresponding dns name
+		// resolver object without any collision.
+		maxRetry := objDetails.collisionCount + 10
+		for ; objDetails.collisionCount < maxRetry; objDetails.collisionCount++ {
+			objDetails.objName = "dns-" + computeHash(dnsName, objDetails.collisionCount)
+			// Check if the generated object name matches with any of the existing
+			// object's name.
+			if _, found := c.objDNSName[objDetails.objName]; !found {
+				break
+			}
+		}
+		// If the maximum retry is reached, then return an error so that it can be
+		// retried again via the retry framework.
+		if objDetails.collisionCount == maxRetry {
```

Comment:
The reason why the object info is added to the map is that it contains the last value of the `collisionCount` and it will be used in the next retry of the reconcile.